src/sentry/utils/arroyo_router.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+import time
+from collections import deque
+from collections.abc import Callable, Mapping, MutableMapping, Sequence
+from typing import Any, Deque, Generic, TypeVar
+
+from arroyo.processing.strategies import MessageRejected
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Message, Partition, TStrategyPayload
+
+from sentry.utils.arroyo_guard import NonFinalStrategy, guard
+
+TResult = TypeVar("TResult")
+
+
+class MergeRoutes(ProcessingStrategy[TResult], Generic[TResult]):
+    def __init__(self, routing_key: str, buffer: MessageBuffer[TResult]) -> None:
+        self.routing_key = routing_key
+        self.buffer = buffer
+
+    def submit(self, message: Message[TResult]) -> None:
+        self.buffer.add(message, self.routing_key)
+
+    def poll(self) -> None:
+        pass
+
+    def join(self, timeout: float | None = None) -> None:
+        pass
+
+    def close(self) -> None:
+        pass
+
+    def terminate(self) -> None:
+        pass
+
+
+@guard
+class Router(
+    NonFinalStrategy[TStrategyPayload, TResult],
+):
+    def __init__(
+        self,
+        next_step: ProcessingStrategy[TResult],
+        route_builders: Mapping[
+            str, Callable[[MergeRoutes[TResult]], ProcessingStrategy[TStrategyPayload]]
+        ],
+        routing_func: Callable[[Message[TStrategyPayload]], str],
+    ) -> None:
+        self.buffer: MessageBuffer[TResult] = MessageBuffer(list(route_builders.keys()))
+        self.routing_func = routing_func
+        self.next_step = next_step
+
+        self.routes: Mapping[str, ProcessingStrategy[TStrategyPayload]] = {
+            routing_key: builder(MergeRoutes(routing_key, self.buffer))
+            for (routing_key, builder) in route_builders.items()
+        }
+
+    def submit(self, message: Message[TStrategyPayload]) -> None:
+        if len(self.buffer) > 10000:
+            raise MessageRejected
+
+        message_route = self.routing_func(message)
+        self.routes[message_route].submit(message)
+        self.buffer.remove(message, message_route)
+
+        while True:
+            next_message = self.buffer.poll()
+            if next_message:
+                self.next_step.submit(next_message)  # this is fine because of backpresure_guard
```

Comment:
- currently the messages are just in the order they finished processing by the routes, and are not reordered by offsets.
- guaranteeing ordering by offsets is impossible because you can always have one route with `{partition1: offset2, partition2: offset1}` and another route with `{partition1: offset1, partition2:offset2}` and there is no valid total order
- so instead all we do is hold onto the message until it's committable offset is below the last processed one on all routes
- so the batch can occur before or after the individual message but we only process it after the individual message has definitely finished processing, and vice versa